Add layout: sidebar and app header

david-mears-2 opened the pull request:
## NB: See [this comment](https://github.com/j-idea/daedalus-web-app/pull/2#issuecomment-2225919107) for how to make the PR diff readable

### PR description

This PR adds what is common between all pages: an openable/closable sidebar, and an app header. These should work on all kinds of devices.

You can even test it directly on your mobile device using your local network by following the instructions in the README:
```bash
npm run dev -- --host
```

It also introduces github action workflows for testing and linting.

There aren't any component tests for the BreadCrumb component because I'm not yet sure how that will be implemented - treat it as an aspect of styling for now. (I created a separate ticket for that.)

### Pages that exist

The blank 'pages' you'll be able to visit are determined by what's in the `pages/` directory, that is:
* localhost:3000/comparison
* localhost:3000/ -- for pages/index.vue

david-mears-2 commented:
I reverted out the 'do linting' commit because it completely clutters up the PR diff. I can add it back after reviews, and then the 'lint' workflow will pass, as shown by the green tick on that commit.

M-Kusumgar reviewed (commented):
Looks good testing on mobile with that QR code is super handy as well! Couple points:
* e2e test was flaky on local, is this something you have experienced? it was the opening and closing sidebar test, kept failing on random browsers every time
* how do i get to comparison page by clicking buttons, im assuming this will happen when you create a couple of scenarios ( so in the future)?

I had a couple questions too!

david-mears-2 commented on components/AppHeader.vue:
```diff
@@ -0,0 +1,87 @@
+const showBreadcrumbs = computed(() => route.meta.hideBreadcrumbs !== true);
+
+const emit = defineEmits(['toggle-sidebar']);
+const toggleSidebar = () => {
+  emit('toggle-sidebar');
+};
```
Yeah. I think this way, either Vue or Nuxt will perform a check that the event emitted is actually explicitly allowed by the defineEmits, but if it comes directly from the template, it doesn't do that check.

david-mears-2 commented on components/AppHeader.vue:
```diff
@@ -0,0 +1,87 @@
+        <CNavItem class="py-1">
+          <VerticalRule/>
+        </CNavItem>
+        <CNavItem href="#"> ?</CNavItem>
```
I don't know how it's going missing for you - it should always be present. These are the buttons in the top right of the app header.

M-Kusumgar replied:
hmmm thats odd then 
![image](https://github.com/user-attachments/assets/0feb9a45-8354-4690-ac3e-77b8b097112f)
can confirm no buttons, are these always present for you? they appear when the screen has smaller width

david-mears-2 replied:
Turns out I just wasn't making my screen wide enough! I'll fix it

david-mears-2 commented on pages/scenario/new.vue:
```diff
@@ -0,0 +1,15 @@
+<template>
+  <div>
+    <Globe />
```
Yes, or more precisely I'm not including it in the PR, as it's WIP. I should have commented this out.

david-mears-2 commented:
> how do i get to comparison page by clicking buttons, im assuming this will happen when you create a couple of scenarios ( so in the future)?

Yes, there's currently no way to get to a comparison by clicking around. You'll be able to get to it from either a single scenario (which you want to compare), or from your bookmarks.

david-mears-2 commented:
> e2e test was flaky on local, is this something you have experienced? it was the opening and closing sidebar test, kept failing on random browsers every time

Yes it has some flakiness for me locally :( but not as bad as 'every time'. I've just run `npm run test:e2e` 5 times in a row and got no failures/flakys.

M-Kusumgar commented:
> Yes it has some flakiness for me locally :( but not as bad as 'every time'. I've just run `npm run test:e2e` 5 times in a row and got no failures/flakys.

maybe its on my end then, ran it 3 times in a row, got multiple failures on every try

absternator commented on components/AppHeader.vue:
```diff
@@ -0,0 +1,87 @@
+import throttle from 'lodash.throttle';
+import { useRoute } from 'vue-router';
+
+const route = useRoute();
```
just a tiny layout thing but may be best to seperate your variables at the top and functions below so there is a clear distinction

david-mears-2 replied:
Hopefully this will be sorted out by https://github.com/j-idea/daedalus-web-app/pull/2#issuecomment-2225919107

absternator commented on components/WebsocketConnection.client.vue:
is there tests for this file?

absternator commented on layouts/default.vue:
```diff
@@ -14,10 +20,30 @@
+}
+
+const handleSidebarHidden = () => {
+  // The 'hide' event is emitted by the CoreUI Sidebar component as a callback whenever the sidebar
```
dont need comment. pretty clean as you did it.. comments just add more noise

absternator reviewed (commented):
Great work looks amazing!! and slick UI!!! and code lgtm too..  just a few minor comments

- maybe a way to pin the sidebar so full size so it dosent auto collapse? or dependent on size of screen? 
- Following on from above, hide makes more sense to go in collapsed mode and not fully gone? Then on smaller screens bring in hamburger menu
- Breadcrumb styling a bit weird. unsure if spacing/styling or im used to it being below
- Would be great to add formatter asap too.. a .prettierrc file should be suffice... wil save you having to go and format everything later
- didnt notice any tests for socket io? is that another PR?

david-mears-2 commented:
> Would be great to add formatter asap too.. a .prettierrc file should be suffice... wil save you having to go and format everything later

It's using [ESLint Stylistic](https://eslint.style/), via `@antfu/eslint-config` -- see README.md

david-mears-2 commented on components/AppHeader.vue:
```diff
@@ -0,0 +1,87 @@
+  window.removeEventListener('scroll', handleScroll);
+});
+
+const VerticalRule = h('div', { class: "vr h-100 mx-2 text-body text-opacity-75" })
```
This was to avoid duplication, but actually since I'm no longer reusing it, I can just go back to html

M-Kusumgar commented on components/BreadCrumb.vue:
```diff
@@ -0,0 +1,25 @@
+import { useRoute } from 'vue-router'
+const route = useRoute()
+
+const items = route.meta.breadcrumbs || []
```
yh might be worth parsing the route in a breadcrumbs component or something like that, probably for another pr though!

david-mears-2 commented:
I worked out an alternative way the sidebar can operate based on @absternator 's suggestion: https://github.com/j-idea/daedalus-web-app/pull/4

david-mears-2 commented:
> Breadcrumb styling a bit weird. unsure if spacing/styling or im used to it being below

What do you mean by 'below'?

david-mears-2 commented:
> didnt notice any tests for socket io? is that another PR?

I will probably only add tests for websockets once we start using them for something. The websockets-related code in this PR is all either setup (copied from docs), or a debugging tool that just proves to the developer that websockets is working.

EmmaLRussell commented:
> I worked out an alternative way the sidebar can operate based on @absternator 's suggestion: #4

Yeah, new way definitely seems preferable. Shouldn't need the hamburger icon on larger screens though, since collapsed view always visible?

david-mears-2 commented:
> Shouldn't need the hamburger icon on larger screens though, since collapsed view always visible?

In the new way, I made it so the hamburger toggles the width of the sidebar - it switches between:
* Fixed open sidebar
and
* Sidebar that opens on hover

So it's about whether we think people will want to be able to fix it open.

EmmaLRussell commented:
> So it's about whether we think people will want to be able to fix it open.

Oh I see - well, they might do, but I'm not sure the hamburger icon is the way to offer that functionality, since it's customary use is to open. Something like a lock icon maybe, but probably something on the sidebar (like your original Hide control) would have been my first thought.

EmmaLRussell commented on components/AppHeader.vue:
```diff
@@ -0,0 +1,87 @@
+})
+</script>
+
+<template>
```
I've always preferred putting the template above the script, as it feels more natural to see your static markup before the dynamic operations which work upon it. But again, this is quite subjectve!

david-mears-2 replied:
I agree actually, so I'm now overriding the lint/format rule that caused this.

david-mears-2 commented:
I pushed a new commit to https://github.com/j-idea/daedalus-web-app/pull/4 that removes the toggle for large screens.

EmmaLRussell commented on pages/comparison.vue:
```diff
@@ -0,0 +1,21 @@
+// A non-dynamic placeholder example of what the breadcrumb text would be.
+// The text should actually depend on the route (which determines which axis is the axis of comparison).
+definePageMeta({
+  breadcrumbs: [
+    { text: 'Base scenario', href: '#' },
+    { text: 'Explore more scenarios', href: '#' },
+    { text: `By policy response` }
+  ]
+})
```
Yeah, it feels like the router ought to be giving us breadcrumb information.

EmmaLRussell commented on pages/index.vue:
```diff
@@ -3,3 +3,12 @@
+
+<script setup lang="ts">
+definePageMeta({
+  hideBreadcrumbs: true
```
..again, it feels like the breadcrumb component itself might make decision about hiding breadbrumb if only one level of route etc, rather than page defining it in the meta.

EmmaLRussell reviewed (commented):
Looks great - really slick. 

I'm not completely convinced by having the logos appear in the sidebar, that Imperial "I" is quite distracting... but I guess they have to go somewhere. We put them on the About page in arbomap - will you need one of those too? - so maybe could go there if acceptable?

The ? is going to be the Help link? It would look more consistent to make that a logo too I think, and give it a tooltip. 

I think the breadcrumbs would look a little cleaner without underline for links, and without the grey horizontal line between header and breadcrumb on smaller screen. But that's very subjective...

Unit tests for components going in a separate ticket?

david-mears-2 commented:
> I'm not completely convinced by having the logos appear in the sidebar, that Imperial "I" is quite distracting... but I guess they have to go somewhere. We put them on the About page in arbomap - will you need one of those too? - so maybe could go there if acceptable?

I'm not sure if we will have an 'About' page in so many words. We'll have a page that explains methodology and links to the paper, which is similar, but might not get called 'About' and I'm not sure the logo/acknowledgements should go on that page.

I've been led to understand that branding it with the name Jameel is important. For now, I will remove the Imperial I so the logo only shows when the sidebar is open (I wasn't convinced about my solution for the distractingness of the I - I was reducing its opacity).

david-mears-2 commented:
> The ? is going to be the Help link? It would look more consistent to make that a logo too I think, and give it a tooltip.

Agreed. I didn't because there wasn't such an icon in the library I was using. Tooltip (you mean just the word 'Help' or something?) comes in https://github.com/j-idea/daedalus-web-app/pull/3. I'm going to add an svg from font awesome to assets.

david-mears-2 commented:
I think I've addressed all comments. I merged into this branch the new way for sidebars to work.

I intend to work out how we 'implement breadcrumbs properly' in [this ticket](https://mrc-ide.myjetbrains.com/youtrack/issue/JIDEA-32/Implement-bread-crumbs-properly).

After final reviews, I'll run `lint:fix` which will do all the formatting and linting, but I won't do that for now because it will make things unreadable.

EmmaLRussell reviewed (commented):
Looks good! Think I preferred the breadcrumb left/center rather than on the right.  

When you enable the formatter/linter, can you make it consistent with our other web apps?
i.e. double quotes, semicolon line endings. We're not completely consistent on indents - I prefer 4 spaces, but we also have 2 spaces...

david-mears-2 commented on tests/components/SideBar.spec.ts:
```diff
@@ -0,0 +1,60 @@
+  CIcon: true,
+}
+
+// Reviewers - I wonder if these tests are too brittle
+// (tied closely to the component's implementation / would change too often)
```
Hmm, well my thoughts are that I already have e2e tests on the overall behavior of the sidebar, making my below tests of the implementation somewhat redundant -- except in that unit tests give quicker feedback about regressions (not only that they are quicker to run but also quicker to pinpoint what broke).

EmmaLRussell replied:
I think unit tests are good for fine detailed test work - you might test edge cases that just involve that component in unit. e2e tests should test the main use cases work end to end, and any behaviours that require multiple components to coordinate in complex ways.

absternator reviewed (commented):
Nice work LGTM.. i wont approve as you and emma still ironing things out :smile:

david-mears-2 commented on layouts/default.vue:
```diff
@@ -20,30 +20,74 @@
+  else {
+    // If this is the 'hide' emitted on page load, we un-do it.
+    // This is because the CoreUI Sidebar component emits a 'hide' event on page load, which we
+    // don't want to obey for larger screen sizes.
```
Good point. I've now refactored this, using [defineModel()](https://vuejs.org/guide/components/v-model)

EmmaLRussell commented:
Code looking good to me now - did you want to fix up the lint build before merging?
